### ISSUE TITLE
Can we get rid of manual angular.module?

### ISSUE THREAD

**Issue by reppners** [ngParty/ng-metadata#81]
I recognized in the examples that there is still the need to declare angular modules and register all the things in the modules manually. Is there a way to do this implicitly?

My team has a codebase which already relies on angular-decorators and now we want to convert it a step closer to ng-2. One of our biggest maintenance burdens are those angular modules (apart from DI but this is solved in ng-metadata) which require quite an amount of manual work while everything else is done by the IDE.

I guess there must be reasons for keeping the modules but I really would like to get rid of 'em in favor of having just ES2015 modules and letting the decorators do the work.

**Comment by Hotell** [comment]
good question! answer is following:

**TL;DR:** 
We can't, because `ngUpgrade` / `unit testing` / `style guides`

**Long version**

- angular 1 has just one global injector, angular 2 has many injector ( hierarchical ) so it would be confusing because it won't be doing what you thought it's doing in angular 1 context
- for unit testing, when you are testing a DOM render of your component, you need to provide module for angular $compiler where that particular component is registered with ( this could be done with some "generator" behind the scenes, but it's not really worth it => possible problems during debugging your unit tests etc... )
- ngUpgrade needs `angular.module` so it knows what and how to upgrade/downgrade in Angular1/2 context
- maybe this will sound strange, but using `angular.modules` and exporting just that module name reference from `index.ts` is following [Angular 2 style guide and separation of concerns principle](https://angular.io/styleguide#!#create-and-import-barrels)

```
src/
|-- app/
|-- |-- dashboard/
|-- |-- |-- dashboard.component.ts|html|css|spec.ts
|-- |-- |-- ...
|-- |-- |-- index.ts
|-- |-- heroes/
|-- |-- |-- ...
|-- |-- |-- index.ts
|-- |-- index.ts
|-- |-- app.component.ts|html|css|spec.ts
|-- |-- main.ts
|-- index.html
```

where every per feature `index.ts` should look like:
```typescript
// src/app/dashboard/index.ts

import * as angular from 'angular';
import { provide } from 'ng-metadata/core';
import { DashboardComponent } from './dashboard.component';

export const DashboardModule = angular.module('dashboard',[])
  .directive(...provide(DashboardComponent))
  .name;
```

and root `index.ts`
```typescript
// src/app/index.ts

import * as angular from 'angular';
import { provide } from 'ng-metadata/core';
import { DashboardModule } from './dashboard';

export const AppModule = angular.module('app',[DashboardModule])
  .directive(...provide(AppComponent))
  .name;
```

and finally `main.ts`
```typescript
// src/app/main.ts

import { bootstrap } from 'ng-metadata/platform';
import { AppModule } from './index';

bootstrap(AppModule);
```

and those indexes will become just the barrels which will export providers/directives 

example:
```typescript
// src/app/dashboard/index.ts

export { DashboardComponent } from './dashboard.component';
```

and so on...

I highly recommend to follow this architecture, I know it's repetitive but you can create live templates in your IDE/Editor  for that :)

also you can [check our scaffold for that] (https://github.com/ngParty/Angular1-scaffold) which also uses [Codelyzer](https://github.com/mgechev/codelyzer)

hope this make things more clear

thanks

**Comment by reppners** [comment]
Yes, this makes the need quite clear. Thx for the comprehensive answer. 👍

**Comment by Hotell** [comment]
my pleasure! pls join our slack group https://ngparty.herokuapp.com/ especially `#ng-metadata` channel and don't hesitate to ask anything out there

cheers!

**Comment by Hotell** [comment]
@reppners  I've created snippets to mitigate the mundane typing hell. check it out
https://github.com/ngParty/jetbrains-angular1-snippets

**Comment by reppners** [comment]
Awesome! Looking forward to use those.
